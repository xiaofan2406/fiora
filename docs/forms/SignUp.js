/* @flow */
import * as React from 'react';
import { signUp } from 'utils/api';
import {
  usernameValidation,
  passwordValidation,
  passwordRepeatValidation,
} from 'utils/validations';
import { Input } from 'widgets';
import createForm from 'fiora/Fiora';

const { Form, Field } = createForm();

const signUpSubmit = async data => {
  const res = await signUp(data);
  if (res.status === 200) {
    console.log('Sign up successful!');
    return null;
  }
  console.log(res.errors);
  return res.errors;
};

const SignUp = () => (
  <Form onSubmit={signUpSubmit} data-testid="signUpForm">
    <Field name="username" onValidate={usernameValidation}>
      {({
        value,
        error,
        isTouched,
        isValidating,
        handleChange,
        handleValidate,
      }) => (
        <Input
          data-testid="usernameInput"
          label="Username"
          value={value}
          error={error}
          isTouched={isTouched}
          isValidating={isValidating}
          onChange={event => {
            handleChange(event.target.value);
            handleValidate();
          }}
        />
      )}
    </Field>
    <Field name="password" onValidate={passwordValidation}>
      {({
        value,
        error,
        isTouched,
        isValidating,
        handleChange,
        handleValidate,
      }) => (
        <Input
          data-testid="passwordInput"
          label="Password"
          value={value}
          error={error}
          isTouched={isTouched}
          isValidating={isValidating}
          onChange={event => {
            handleChange(event.target.value);
          }}
          onBlur={handleValidate}
        />
      )}
    </Field>
    <Field name="passwordRepeat" onValidate={passwordRepeatValidation}>
      {({
        value,
        error,
        isTouched,
        isValidating,
        handleChange,
        handleValidate,
      }) => (
        <Input
          data-testid="passwordRepeatInput"
          label="Confirm password"
          value={value}
          error={error}
          isTouched={isTouched}
          isValidating={isValidating}
          onChange={event => {
            handleChange(event.target.value);
          }}
          onBlur={handleValidate}
        />
      )}
    </Field>
    <button type="submit">submit</button>
  </Form>
);

export default SignUp;
