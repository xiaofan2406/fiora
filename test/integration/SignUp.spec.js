import React from 'react';
import { render, fireEvent, waitForElement } from 'react-testing-library';
import SignUp from './SignUp';

it('matches snapshot', () => {
  const { container } = render(<SignUp />);
  expect(container.firstChild).toMatchSnapshot();
});

it('has 3 inputs with not initial values', () => {
  const { getByTestId } = render(<SignUp />);
  expect(getByTestId('usernameInput').value).toBe('');
  expect(getByTestId('passwordInput').value).toBe('');
  expect(getByTestId('passwordRepeatInput').value).toBe('');
});

it('updates value when input onChange is triggered', () => {
  const { getByTestId } = render(<SignUp />);

  const usernameInput = getByTestId('usernameInput');
  const passwordInput = getByTestId('passwordInput');
  const passwordRepeatInput = getByTestId('passwordRepeatInput');
  fireEvent.change(usernameInput, { target: { value: 'new username' } });
  fireEvent.change(passwordInput, { target: { value: '1234567' } });
  fireEvent.change(passwordRepeatInput, { target: { value: '1' } });

  expect(usernameInput.value).toBe('new username');
  expect(passwordInput.value).toBe('1234567');
  expect(passwordRepeatInput.value).toBe('1');
});

describe('username input', () => {
  it('does not have any error initially', () => {
    const { getByTestId } = render(<SignUp />);
    try {
      getByTestId('usernameError');
    } catch (err) {
      expect(err.message).toMatch(
        'Unable to find an element by: [data-testid="usernameError"]'
      );
    }
  });

  it('validates username whenever the input is changed', async () => {
    const { getByTestId, getByText } = render(<SignUp />);

    fireEvent.change(getByTestId('usernameInput'), { target: { value: 'n' } });
    await waitForElement(() => getByTestId('usernameError'));

    expect(getByText(/^error:/).innerHTML).toMatch(
      'Username should be as least 5 characters'
    );

    fireEvent.change(getByTestId('usernameInput'), {
      target: { value: 'new username' },
    });

    try {
      getByTestId('usernameError');
    } catch (err) {
      expect(err.message).toMatch(
        'Unable to find an element by: [data-testid="usernameError"]'
      );
    }
  });

  it('validates username whenever the input is changed', async () => {
    const { getByTestId, getByText } = render(<SignUp />);

    fireEvent.change(getByTestId('usernameInput'), { target: { value: 'n' } });
    await waitForElement(() => getByTestId('usernameError'));

    expect(getByText(/^error:/).innerHTML).toMatch(
      'Username should be as least 5 characters'
    );
  });
});
