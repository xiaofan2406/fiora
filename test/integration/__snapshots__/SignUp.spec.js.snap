// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  data-testid="signUpForm"
>
  <input
    data-testid="usernameInput"
    value=""
  />
  <input
    data-testid="passwordInput"
    value=""
  />
  <input
    data-testid="passwordRepeatInput"
    value=""
  />
  <button
    type="reset"
  >
    Reset
  </button>
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
